replicated PMM2 improvements to HA (#3511)
percona/pmm: documentation/docs/install-pmm/HA.md

- Consider your specific uptime needs, performance requirements, and potential data loss you can tolerate, while also keeping in mind PMM's limitations.

### 1. Simple Docker restart with data caching

The most straightforward approach to increase availability in PMM is to launch the PMM Server within Docker using the `--restart=always` flag. See [Install PMM Server with Docker](../install-pmm/install-pmm-server/baremetal/docker/index.md) for more information.

This ensures that the PMM Server automatically restarts if a minor issue occurs. Additionally, PMM's data caching feature stores data locally on the PMM Client when the connection to the PMM Server is interrupted.

Once the connection is restored, the cached data is transferred to the PMM Server, ensuring no data loss during the restart process.

This option is suitable for scenarios where the primary concern is the ability to investigate potential issues later. However, it's important to note that this approach is limited by the underlying physical infrastructure. If the failure stems from a hardware issue, automatic recovery might be challenging.

### 2. Leverage Kubernetes for enhanced isolation

> If you are running PMM in a Kubernetes (K8s) environment, PMM offers a Helm chart that facilitates running PMM with enhanced isolation. See [Install PMM Server with Docker](../install-pmm/install-pmm-server/baremetal/docker/index.md).

Reviewer: ```suggestion
If you are running PMM in a Kubernetes (K8s) environment, PMM offers a Helm chart that facilitates running PMM with enhanced isolation. See [Install PMM Server with Helm on the Kubernetes clusters](install-pmm-server/baremetal/helm/index.html).
```

In this setup, even if the physical infrastructure encounters a problem, K8s automatically handles failover, migrating the PMM instance to a healthy node. 

While restarts within K8s can take up to several minutes (depending on your infrastructure configuration), PMM's data caching ensures that information is preserved during this transition. Alerts will still be triggered to keep you informed about any issues that started during PMM's restart and continue after PMM is back.

### 3. Fully-clustered PMM in Kubernetes (in development)

> If you have a large deployment with numerous instances and distributed locations, you might find that a fully clustered PMM setup in Kubernetes is better suited to your needs. We are actively developing this solution, which is slated for release in Q3/2024, to cater specifically to users managing extensive and complex monitoring environments.

Reviewer 2: ```suggestion
If you have a large deployment with numerous instances and distributed locations, you might find that a fully clustered PMM setup in Kubernetes is better suited to your needs. We are actively developing this solution, which is slated for release later with PMM 3.x, to cater specifically to users managing extensive and complex monitoring environments.
```

        clickhouse/clickhouse-server:23.8.2.7-alpine
        ```
    

>     === "Run services on a seperate instance"

Reviewer 3: should be "separate"

        --hostname ${PMM_ACTIVE_NODE_ID} \
        --network pmm-network \
        --ip ${PMM_ACTIVE_IP} \
        -e PMM_DISABLE_BUILTIN_CLICKHOUSE=1 \
        -e PMM_DISABLE_BUILTIN_POSTGRES=1 \
        -e PMM_CLICKHOUSE_ADDR=${CH_HOST_IP}:9000 \
        -e PMM_CLICKHOUSE_DATABASE=pmm \
        -e PMM_CLICKHOUSE_BLOCK_SIZE=10000 \
        -e PMM_CLICKHOUSE_POOL_SIZE=2 \

>         -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \
>         -e PMM_POSTGRESUSERNAME=${PG_USERNAME} \
>         -e PMM_POSTGRESDBPASSWORD=${PG_PASSWORD} \

Reviewer 2: ```suggestion
        -e PMM_POSTGRES_ADDR=${PG_HOST_IP}:5432 \
        -e PMM_POSTGRES_USERNAME=${PG_USERNAME} \
        -e PMM_POSTGRES_DBPASSWORD=${PG_PASSWORD} \
```

        ${PMM_DOCKER_IMAGE}
        ```
    
    === "Run services on a seperate instance"
    
        ```sh
        docker run -d \
        --name ${PMM_ACTIVE_NODE_ID} \
        -p 8080:8080 \
        -p 8443:8443 \
        -p 9094:9094 \
        -p 9096:9096 \
        -p 9094:9094/udp \
        -p 9096:9096/udp \
        -p 9097:9097 \
        -e PMM_DISABLE_BUILTIN_CLICKHOUSE=1 \
        -e PMM_DISABLE_BUILTIN_POSTGRES=1 \
        -e PMM_CLICKHOUSE_ADDR=${CH_HOST_IP}:9000 \
        -e PMM_CLICKHOUSE_DATABASE=pmm \
        -e PMM_CLICKHOUSE_BLOCK_SIZE=10000 \
        -e PMM_CLICKHOUSE_POOL_SIZE=2 \

>         -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \

Reviewer 3: should be PMM_POSTGRES_ADDR

>         -e PMM_POSTGRESUSERNAME=${PG_USERNAME} \

Reviewer 3: PMM_POSTGRES_USERNAME

>         -e PMM_POSTGRESDBPASSWORD=${PG_PASSWORD} \

Reviewer 3: PMM_POSTGRES_DBPASSWORD

        --hostname ${PMM_PASSIVE_NODE_ID} \
        --network pmm-network \
        --ip ${PMM_PASSIVE_IP} \
        -e PMM_DISABLE_BUILTIN_CLICKHOUSE=1 \
        -e PMM_DISABLE_BUILTIN_POSTGRES=1 \
        -e PMM_CLICKHOUSE_ADDR=${CH_HOST_IP}:9000 \
        -e PMM_CLICKHOUSE_DATABASE=pmm \
        -e PMM_CLICKHOUSE_BLOCK_SIZE=10000 \
        -e PMM_CLICKHOUSE_POOL_SIZE=2 \

>         -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \

Reviewer 3: same for Postgres variables

>         -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \
>         -e PMM_POSTGRESUSERNAME=${PG_USERNAME} \
>         -e PMM_POSTGRESDBPASSWORD=${PG_PASSWORD} \

Reviewer 2: ```suggestion
        -e PMM_POSTGRES_ADDR=${PG_HOST_IP}:5432 \
        -e PMM_POSTGRES_USERNAME=${PG_USERNAME} \
        -e PMM_POSTGRES_DBPASSWORD=${PG_PASSWORD} \
```

        ${PMM_DOCKER_IMAGE}
        ```
    
    === "Run services on a seperate instance"
    
        ```sh
        docker run -d \
        --name ${PMM_PASSIVE_NODE_ID} \
        -p 8080:8080 \
        -p 8443:8443 \
        -p 9094:9094 \
        -p 9096:9096 \
        -p 9094:9094/udp \
        -p 9096:9096/udp \
        -p 9097:9097 \
        -e PMM_DISABLE_BUILTIN_CLICKHOUSE=1 \
        -e PMM_DISABLE_BUILTIN_POSTGRES=1 \
        -e PMM_CLICKHOUSE_ADDR=${CH_HOST_IP}:9000 \
        -e PMM_CLICKHOUSE_DATABASE=pmm \
        -e PMM_CLICKHOUSE_BLOCK_SIZE=10000 \
        -e PMM_CLICKHOUSE_POOL_SIZE=2 \
        -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \

>         -e PMM_POSTGRESUSERNAME=${PG_USERNAME} \

Reviewer 3: again variables for Postgres

>         -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \
>         -e PMM_POSTGRESUSERNAME=${PG_USERNAME} \
>         -e PMM_POSTGRESDBPASSWORD=${PG_PASSWORD} \

Reviewer 2: ```suggestion
        -e PMM_POSTGRES_ADDR=${PG_HOST_IP}:5432 \
        -e PMM_POSTGRES_USERNAME=${PG_USERNAME} \
        -e PMM_POSTGRES_DBPASSWORD=${PG_PASSWORD} \
```

        ${PMM_DOCKER_IMAGE}
        ```
    

>     === "Run services on a seperate instance"

Reviewer 3: separate

    
        ```sh
        docker run -d \
        --name ${PMM_PASSIVE2_NODE_ID} \
        -p 80:8080 \
        -p 443:8443 \
        -p 9094:9094 \
        -p 9096:9096 \
        -p 9094:9094/udp \
        -p 9096:9096/udp \
        -p 9097:9097 \
        -e PMM_DISABLE_BUILTIN_CLICKHOUSE=1 \
        -e PMM_DISABLE_BUILTIN_POSTGRES=1 \
        -e PMM_CLICKHOUSE_ADDR=${CH_HOST_IP}:9000 \
        -e PMM_CLICKHOUSE_DATABASE=pmm \
        -e PMM_CLICKHOUSE_BLOCK_SIZE=10000 \
        -e PMM_CLICKHOUSE_POOL_SIZE=2 \

>         -e PMM_POSTGRESADDR=${PG_HOST_IP}:5432 \
>         -e PMM_POSTGRESUSERNAME=${PG_USERNAME} \
>         -e PMM_POSTGRESDBPASSWORD=${PG_PASSWORD} \

Reviewer 2: ```suggestion
        -e PMM_POSTGRES_ADDR=${PG_HOST_IP}:5432 \
        -e PMM_POSTGRES_USERNAME=${PG_USERNAME} \
        -e PMM_POSTGRES_DBPASSWORD=${PG_PASSWORD} \
```
